### PR TITLE
feat(discover2): Add draggable columns on discover2 table

### PR DESCRIPTION
**Pull request by dashed**
## TODO

- [x] defer to Danny regarding referencing `.grid-head-cell-button`; or at least the right approach. EDIT: see: https://github.com/getsentry/sentry/pull/14980/commits/4bda109de193db22c4087df699404af47877e59a
- [x] clean up things
- [x] more emphasis on the column being dragged
- [x] do not reload the table.

------

Closes SEN-1126

### CODE REVIEW

File: src/sentry/static/sentry/app/components/gridEditable/flags.tsx [hunk 1]
```diff
@@ -7,4 +7,4 @@
  * 3. Allow drag-to-rearrage columns
  */
 export const FLAG_GRID_RESIZABLE = false;
-export const FLAG_GRID_DRAGGABLE = false;
+export const FLAG_GRID_DRAGGABLE = true;
```
Comment (dashed) [inline src/sentry/static/sentry/app/components/gridEditable/flags.tsx]: @dashed remember to revert this change

Comment (markstory) [reply src/sentry/static/sentry/app/components/gridEditable/flags.tsx]: Shouldn't this flag be removed once dragging is working?

File: src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+    event: React.MouseEvent<SVGSVGElement, MouseEvent>,
+    initialColumnIndex: number
+  ) => {
+    const isDragging = this.state.isDragging;
+
+    if (isDragging || event.type !== 'mousedown') {
+      return;
+    }
+
+    // prevent the user from selecting things outside the minimap when dragging
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx]: Do you mean outside the column header group?

File: src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+      return column.getBoundingClientRect();
+    });
+
+    console.log('foo', foo);
+    console.log('moo', moo);
+
+    foo.forEach(element => {
+      const rects = element.getBoundingClientRect();
+
+      const reference = document.createElement('div');
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx]: Would this be a good place to use a portal to render the ghost?

File: src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    const childrenProps = {
+      startColumnDrag: this.startColumnDrag,
+      draggingColumnIndex: this.state.draggingColumnIndex,
+      destinationColumnIndex: this.state.destinationColumnIndex,
+    };
+
+    return this.props.children(childrenProps);
+  };
+
+  renderThing = () => {
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx]: Could call this 'activeDragPlaceholder'

File: src/sentry/static/sentry/app/components/gridEditable/flags.tsx (hunk 1 above)
Comment (dashed) [reply src/sentry/static/sentry/app/components/gridEditable/flags.tsx]: I'm going to check with @leedongwei ; I assumed this was a feature-flagged feature.

Comment (dashed) [reply src/sentry/static/sentry/app/components/gridEditable/flags.tsx]: @leedongwei gotcha. Thanks for clarifying 👍

File: src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx (hunk 3 above)
Comment (dashed) [reply src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx]: These were DOM elements I've created for debugging; I've since removed them.

File: src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx (hunk 4 above)
Comment (dashed) [reply src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx]: renamed.

File: src/sentry/static/sentry/app/views/eventsV2/table/tableView.tsx [hunk 5]
```diff
@@ -119,22 +121,23 @@ class TableView extends React.Component<TableViewProps> {
...
     const {eventView, location, tableData} = this.props;
+
     if (!tableData) {
       return column.name;
     }
 
-    const field = eventView.fields[columnIndex];
+    const field = column.eventViewField;
+
+    // establish alignment based on the type
```
Comment (leedongwei) [inline src/sentry/static/sentry/app/views/eventsV2/table/tableView.tsx]: Nit: This blank line can be removed.

File: src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    initialColumnIndex: number
+  ) => {
+    const isDragging = this.state.isDragging;
+
+    if (isDragging || event.type !== 'mousedown') {
+      return;
+    }
+
+    // prevent the user from selecting things when dragging a column.
+
```
Comment (leedongwei) [inline src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx]: Nit: Removable blank line.

Comment (leedongwei) [reply src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx]: There's a few more below but I'll only mark this one.

File: src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+    }
+
+    // indicate drag has ended
+
+    const destinationColumnIndex = this.state.destinationColumnIndex;
+    const draggingColumnIndex = this.state.draggingColumnIndex;
+
+    this.setState({
+      isDragging: false,
+      left: void 0,
```
Comment (leedongwei) [inline src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx]: Would it be better to use `undefined`?

**Review by leedongwei — COMMENTED** [review]
Looks really good! 

🙌

File: src/sentry/static/sentry/app/components/gridEditable/styles.tsx [hunk 8]
```diff
@@ -116,7 +117,17 @@ export const GridHeadCellButton = styled('div')<GridEditableProps>`
   text-overflow: ellipsis;
   overflow: hidden;
 
-  background: ${p => (p.isEditing ? p.theme.offWhite2 : 'none')};
+  background: ${p => {
+    if (p.isDragging) {
+      return '#CEC2D8';
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/gridEditable/styles.tsx]: Should this be a theme value or use `color` to derive from a theme value?

File: src/sentry/static/sentry/app/components/gridEditable/styles.tsx [hunk 9]
```diff
@@ -195,9 +206,9 @@ export const GridHeadCellButtonHoverBackground = styled(GridHeadCellButton)`
     color: ${p => p.theme.gray1} !important;
   }
 `;
-export const GridHeadCellButtonHoverButtonGroup = styled('div')<GridEditableProps>`
-  ${p => !p.isFlagged && 'margin: 0 auto;'}
-`;
+
+export const GridHeadCellButtonHoverButtonGroup = styled('div')<GridEditableProps>``;
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/gridEditable/styles.tsx]: If this component is unstyled do we need it?

File: src/sentry/static/sentry/app/views/eventsV2/eventView.tsx [hunk 10]
```diff
@@ -759,6 +760,10 @@ class EventView {
       delete payload.cursor;
     }
 
+    if (!payload.query) {
+      delete payload.query;
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/eventsV2/eventView.tsx]: Why? The tag facets should take the current conditions into account or you lose the ability to drill down into results.

File: src/sentry/static/sentry/app/views/eventsV2/utils.tsx [hunk 11]
```diff
@@ -102,9 +102,11 @@ export function fetchTagDistribution(
   key: string,
   query: EventQuery
 ): Promise<Tag> {
-  const urlParams = pick(query, Object.values(URL_PARAM));
+  // const urlParams = pick(query, Object.values(URL_PARAM));
```
Comment (markstory) [inline src/sentry/static/sentry/app/views/eventsV2/utils.tsx]: Could delete this.

File: src/sentry/static/sentry/app/views/eventsV2/utils.tsx [hunk 12]
```diff
@@ -228,6 +228,8 @@ const TEMPLATE_TABLE_COLUMN: TableColumn<React.ReactText> = {
   name: '',
   aggregation: '',
   field: '',
+  eventViewField: {field: '', title: ''},
```
[the remark this answers is not in the document]
Comment (dashed) [reply src/sentry/static/sentry/app/views/eventsV2/utils.tsx]: I did both. https://github.com/getsentry/sentry/pull/14980/commits/b0e59c072f219cfe645fa4d5263755f54a603aa1 and https://github.com/getsentry/sentry/pull/14980/commits/27abd12af9fc96881d75b3dbaf984a4c3612ec33

File: src/sentry/static/sentry/app/components/gridEditable/styles.tsx (hunk 8 above)
Comment (dashed) [reply src/sentry/static/sentry/app/components/gridEditable/styles.tsx]: Updated in https://github.com/getsentry/sentry/pull/14980/commits/c6a43812578e55fe517681ff2feeeb1a677ec3ae

File: src/sentry/static/sentry/app/views/eventsV2/eventView.tsx (hunk 10 above)
Comment (dashed) [reply src/sentry/static/sentry/app/views/eventsV2/eventView.tsx]: I'm reverting this. I'm going to put a revised payload shape for tags in a separate PR.

File: src/sentry/static/sentry/app/views/eventsV2/utils.tsx (hunk 11 above)
Comment (dashed) [reply src/sentry/static/sentry/app/views/eventsV2/utils.tsx]: Reverting this and putting it in a separate PR.

File: src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx (hunk 7 above)
Comment (dashed) [reply src/sentry/static/sentry/app/views/eventsV2/table/draggableColumns.tsx]: They should mean the same thing. Either isn't better any than the other. 

I generally prefer `void 0` because it's shorter to type 😆 . Uglifyjs does this for the same reasons to trim as much characters as possible.

**Review by markstory — APPROVED** [review]
Looks good from playing with it locally. The visuals and interaction are roughed in and can be refined once we get feedback from others.
